File: haskell/028/Main.hs
{-
    ProjectEuler 028
    Haskell / Main.hs
    copyright (c) 2015 Susisu

    GHC 7.8.3
    `ghc Main.hs -o bin/main`
-}

module Main where

main :: IO ()
main = print $ total (1001 `div` 2)
    where
        total n = 1 + itr n 0 2 1 0 + itr n 0 4 1 0 + itr n 0 6 1 0 + itr n 0 8 1 0
        itr n i d a s
            | i >= n     = s
            | otherwise = let a' = a + d in itr n (i + 1) (d + 8) a' (s + a')

File: javascript/028/main.js
/*
 * ProjectEuler 028
 * JavaScript / main.js
 * copyright (c) 2015 Susisu
 *
 * Node.js v5.0.0
 * `node main.js`
 */

"use strict";

function main() {
    var L = 1001;
    var n = (L + 1) / 2;
    var k = 0;
    var sum = 0;
    for (var i = 0; i < n; i++) {
        k += 1;
        if (i == 0) {
            sum += k;
            continue;
        }
        var l = 2 * i + 1;
        k += l - 2; // SE
        sum += k;
        k += l - 1; // SW
        sum += k;
        k += l - 1; // NW
        sum += k;
        k += l - 1; // NE
        sum += k;
    }
    console.log(sum);
}


main();
